### PR TITLE
RSA Validation of Hart CVRs

### LINKED ISSUE #747: Programmatically verify Hart signed CVR files
Hart has provided us with a set of digitally signed CVRs and the public key necessary to verify them. Can we programmatically verify those files using the public key they have provided. 

For now, in 1.3.2 if we cannot verify the signature it should be a halting exception. This is to satisfy CA requirements. It doesn't need to be that strict for other jurisdictions. We discussed implemented a two build system - one for the stricter CA requirements and one for other jurisdictions. I'll create a separate issue for that.

Contact me directly for the signed files and public key!

### PR DESCRIPTION
**Pull request by artoonie**
resolves #747 

Security flags will be toggled when #761 is merged

### CODE REVIEW

**Comment by tarheel** [comment]
Does WIP mean we should hold off on reviewing this for now?

**Comment by artoonie** [comment]
Yes, hold off for now

**Review by artoonie — COMMENTED** [review]
@tarheel @HEdingfield this is ready for review!

**Review by tarheel — COMMENTED** [review]
Looks good to me! I hope you were able to find code samples for a lot of this vs. having to figure it out on your own. I'll let @HEdingfield look also.

File: src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java
Comment (artoonie) [reply src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java:76]: Agree, the only downside is that we're allowed to throw from anywhere in a function, but we can't have multiple returns, so this would make these functions messier.

I guess I'll summarize the current logic as so:
1. We throw if we were unable to check the signature
2. We return false if the signature didn't match

I agree it's weird. I'm a little more partial to _always_ throwing, since that would make the code cleaner, and the requirements specify it must be a "halting exception," and when we return false we end up throwing anyway.

File: src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java
Comment (artoonie) [reply src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java:120]: It doesn't but the modulus is almost always the same -- required in some standards, suggested by others. AQAB = 65537. See also: https://crypto.stackexchange.com/questions/3110/impacts-of-not-using-rsa-exponent-of-65537

So I opted to not put it to keep it simple here.

File: src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java
Comment (artoonie) [reply src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java:175]: _you have no idea_ 😅

File: src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java
Comment (artoonie) [reply src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java]: I opted to throw two types of exceptions, rather than having one type of error return a boolean and another throw an exception. How's the updated code look?

File: src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java
Comment (tarheel) [inline src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java:63]: The distinction between this two exception types is a bit unclear given the names you chose. Maybe something like `VerificationUnableToRunException` and `VerificationSignatureDidNotMatchException`?

File: src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java
Comment (tarheel) [inline src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java:74]: Yeah, this is better.

**Review by tarheel — APPROVED** [review]
Looks good!

File: src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java
Comment (artoonie) [reply src/main/java/network/brightspots/rcv/SecuritySignatureValidation.java:63]: Agreed.

What I'm trying to get at is that it's possible for the signature to be correct, but we didn't even check the signature because there were other errors first. For example, the signature could be correct, but the key was different; or the signature was correct, but the file hash was different.

For that reason, I'll use `VerificationDidNotRun` and `VerificationSignatureDidNotMatch`

**Pull request merged**
